### PR TITLE
feat: change counts in cart

### PR DESCRIPTION
**Pull request by NickyMeuleman**
ref #118 
Hi 👋 !
Will attempt to add this functionality.
Can you tell me if I'm going in the right direction please?
If I (roughly) am, how should I implement the function that changing the quantity calls?

### CODE REVIEW

File: src/components/Cart/OpenCart.js [hunk 1]
```diff
@@ -81,6 +81,11 @@ export default () => (
         event.preventDefault();
         removeLineItem(client, checkout.id, itemID);
       };
+      const handleQuantityChange = (id, quantity) => event => {
+        console.log('updatequantity');
+
+        // logic to update cart
```
Comment (jlengstorf) [inline src/components/Cart/OpenCart.js]: This would be an update to the line item, which is documented in the [Shopify Buy SDK](https://shopify.github.io/js-buy-sdk/#updating-line-items).

Basically, get the item ID, the checkout ID, and send a call to the `updateLineItems`.

This will be really similar to how `removeLineItem` is implemented here: https://github.com/gatsbyjs/store.gatsbyjs.org/blob/7dd6752073a7812a9d668620633f16f3ac52aba2/src/components/shared/Layout.js#L106-L112

If you follow the implementation of that function in the `OpenCart` component, you can see where all the args are coming from: https://github.com/gatsbyjs/store.gatsbyjs.org/blob/7dd6752073a7812a9d668620633f16f3ac52aba2/src/components/Cart/OpenCart.js#L79-L83

Does that make sense?

**Comment by NickyMeuleman** [comment]
I just spent way too much time hunting down a bug.

It was a typo, I mistyped one letter and read over it multiple times 😅

File: src/components/shared/Layout.js [hunk 2]
```diff
@@ -110,7 +110,17 @@ export default class Layout extends React.Component {
           }));
         });
       },
-      // updateQuantity: (lineItemID, quantity) => {},
+      updateLineItem: (client, checkoutID, lineItemID, quantity) => {
+        const lineItemsToUpdate = [{ id: lineItemID, quantity: parseInt(quantity, 10) }];
```
Comment (NickyMeuleman) [inline src/components/shared/Layout.js]: what gets passed in under lineItemID is actually the variantID. Should change that.

**Comment by jlengstorf** [comment]
@NickyMeuleman is this ready for a review?

**Comment by NickyMeuleman** [comment]
Not yet. No computer atm, will revisit this later today

**Comment by NickyMeuleman** [comment]
@jlengstorf ready for review!

File: src/components/Cart/LineItem.js [hunk 3]
```diff
@@ -77,7 +101,16 @@ export default ({ item, handleRemove }) => (
...
-    <Quantity>{item.quantity}</Quantity>
+    <HiddenLabel htmlFor={`quantity_${item.id.substring(58, 64)}`}>Quantity:</HiddenLabel>
+    <Quantity
+      id={`quantity_${item.id.substring(58, 64)}`}
+      type="number"
+      name="quantity"
+      min="1"
+      step="1"
+      onChange={event => updateQuantity(item.id, event.target.value)}
+      defaultValue={item.quantity}
```
Comment (NickyMeuleman) [inline src/components/Cart/LineItem.js]: Went with `defaultValue` here because it immidiately gives the user feedback as to the number they chose. `value` also works, but you have to wait until the data returns (the props change) before you see an update. That can cause confusion/multiple clicks because the number you see doesn't change immidiately.

File: src/components/Cart/LineItem.js [hunk 4]
```diff
@@ -77,7 +101,16 @@ export default ({ item, handleRemove }) => (
...
     </ItemInfo>
-    <Quantity>{item.quantity}</Quantity>
+    <HiddenLabel htmlFor={`quantity_${item.id.substring(58, 64)}`}>Quantity:</HiddenLabel>
+    <Quantity
+      id={`quantity_${item.id.substring(58, 64)}`}
+      type="number"
+      name="quantity"
+      min="1"
+      step="1"
+      onChange={event => updateQuantity(item.id, event.target.value)}
```
Comment (NickyMeuleman) [inline src/components/Cart/LineItem.js]: Is this ok, or should this be handled differently? (e.g. making LineItem stateful and making this a controlled component)

Comment (NickyMeuleman) [reply src/components/Cart/LineItem.js]: Working on it! Time to flex that Promise muscle (I'm thinking about setting the isLoaded state to true in the .then of the updateLineItem function, which currently is not a Promise)

File: src/components/Cart/LineItem.js [hunk 5]
```diff
@@ -64,26 +92,54 @@ const Remove = styled('a')`
...
+class LineItem extends React.Component {
+  state = {
+    quantity: this.props.item.quantity || 0
+  }
+  inputChangeHandler(event) {
+    const target = event.target
+    const value = target.value
+
+    this.setState({ quantity: value })
+    this.props.updateQuantity(value)
```
Comment (NickyMeuleman) [inline src/components/Cart/LineItem.js]: Should we debounce calls to this function?

File: src/components/Cart/LineItem.js (hunk 4 above)
Comment (jlengstorf) [reply src/components/Cart/LineItem.js]: Cool. You can also potentially `async`/`await` this if a `Promise` chain feels to messy/confusing.

**Comment by NickyMeuleman** [comment]
I accidentally pushed to `upstream` instead of my branch (which also starts with `up`)
sorry! 🙏 
Good news is: finally figured out how to set the loading state earlier while keeping only one call to `updateLineItem`. Just need to figure out an edge-case and format now.

**Comment by NickyMeuleman** [comment]
Done! Ready for another look.

PS: on a related note, are you using the PR extension for VSCode to test this out? I heard good things about it!

**Comment by jlengstorf** [comment]
This looks great, @NickyMeuleman!

I just opened https://github.com/NickyMeuleman/store.gatsbyjs.org/pull/1 with a small change to the loading style, but otherwise this is ready to go, I think.

To answer your questions:
1. That loading animation would be awesome, but is probably too much work for the immediate future. I'd love to see this happen, though, if someone has time for it!
2. I did use the PR extension for VS Code to try it out. It's really handy!

**Comment by jlengstorf** [comment]
Hey, @NickyMeuleman — it looks like we need to merge upstream changes into this branch. Can you merge `master` into your branch so you've got all the latest packages installed?

You _might_ be able to rebase, which would be ideal, but it doesn't really matter either way.

**Comment by NickyMeuleman** [comment]
Ooh, merge conflicts 😅 
I'll merge(with rebase) master.
Resolve the conflicts in `package.json`

For the `yarn.lock`, I'll regenerate it and commit the new version.

Hold while git shows me a bunch of errors 🤣
